Denormalize pass

e-kayrakli opened the pull request:
This PR adds a `denormalize` pass right before codegen. Purpose of the pass is
to remove local variables which are def'd and use'd only once.

DETAILS

- A local variable is considered for denormalization if:
  - It has only one def and use.
  - Its def satisfies the following
    - It's a PRIM_MOVE or PRIM_ASSIGN,
    - PRIM_MOVE/PRIM_ASSIGN cannot create a communication, ie
      - Neither LHS nor RHS is wide.
      - RHS is not a MEMBER primitive with wide base.
    - RHS and LHS of PRIM_MOVE are of same type
    - Neither RHS nor LHS is extern.
  - Its use satisfies the following
    - It's not: PRIM_{ADDR_OF, ARRAY_GET, GET_MEMBER, DEREF, GET_MEMBER_VALUE}
    - If it's a PRIM_MOVE, LHS and RHS are of same type.
    - It is not condStmt in a while family loop or condBlock/incrBlock in
      CForLoop

- Given we found candidates, denormalization of the temporary is aborted if
  there is an unsafe CallExpr between def and use

- --[no-]denormalize flag and CHPL_DENORMALIZE controls the behaviour. Default
  is --no-denormalize. You have to pass -denormalize or set CHPL_DENORMALIZE=1
  to enable denormalization.

- Minor change in dead code elimination pass.

  Previously codegen pass was ignoring some of the primitives unless they are
  RHS of PRIM_MOVE. Same assumption no longer holds. From this category of
  primitives, dead code elimination didn't eliminate PRIM_ARRAY_GET for some
  reason, generating bunch of unused statement warnings.

`CallExpr` SAFETY

A CallExpr is considered safe for reordering if:

- It is a non-essential primitive,
- If it's not a primitive:
  - It doesn't have any ref formals,
  - It doesn't touch(neither def nor use) a global
  - It doesn't touch(neither def nor use) an extern
  - Function body doesn't have any essential primitive

CODEGEN CHANGES

This PR also makes important changes in code generation, mostly confined
to `expr.cpp`. Some codegen changes highlights:

- `codegenPrimMoveRhsIsSpecialPrimitive` was used in `codegenPrimMove`
  only to generate code with some primitives, whose generated code might
  change depending on the LHS of the PRIM_MOVE. Now that the AST can be
  denormalized, it is possible to have those "special" primitives
  elsewhere.

  Therefore, said helper function is transformed into

  ```
  static bool codegenIsSpecialPrimitive(BaseAST* target, Expr* e,
    GenRet& ret);
  ```
  where `target` is the context this primitive is going to be used. If
  it's in a PRIM_MOVE then it is the LHS, if primitive is used as an
  actual then `target` is the corresponding formal. It is OK to pass
  NULL where target is not applicable.

  Function returns if the primitive passed in `e` is handled by the
  function(ie is "special")

- We now generate typecasts for all integer operations. This is due to C
  promoting small integers to full ints in arithmetic operations. When
  such integers are in arithmetic operations and moved to somewhere
  other than a variable assignment to a corresponding type, we have to
  make sure they are typecast(and truncated)

  Currently there is no easy way of getting c_int size for target
  architecture in the compiler. Therefore casts are not limited to small
  ints.

- Minor adjustment in conditional code generation to prevent extra
  parentheses

WHAT CHANGES WITH DENORMALIZE

- Better generated C code vs Chapel code fit.
- FWIW I observed 25-40% decrease in generated C code size. Code decrease is
  generally larger when compiled with --no-local
  - lulesh.chpl goes down from ~160K to ~100K LOC in generated c-code. with
    --no-local it's ~260K to ~170K.
- Changes in compilation time
  - Although generated code is smaller, backend C compiler I have used doesn't
    get significantly faster by this. I generally observe some decrease in
    compilation time but it is not very noteworthy.
- Changes in execution time
  - When no-fast --normalize makes execution time 10-15% faster, I'd speculate
    that's due to backend compiler being too scared to touch temporaries and
    optimize some function calls.
  - When fast, I didn't observe any clear impact of denormalization. As backend
    compiler probably handles that itself with -O3 or whatever the flag.

TEST STATUS

- std config
- std config --no-local
- std config --fast
- std config --denormalize --no-local
- std config --denormalize --fast

PS Currently test/performance/compiler/elliot/passCheck.chpl fails

vasslitvinov commented on compiler/AST/expr.cpp:
```diff
@@ -5858,124 +5885,112 @@ GenRet CallExpr::codegenPrimMove() {
-  CallExpr* rhs    = toCallExpr(get(2));
+static bool codegenIsSpecialPrimitive(BaseAST* target, Expr* e, GenRet& ret) {
   bool      retval = false;
+  CallExpr* ce = toCallExpr(e);
```
I suggest renaming 'ce' to 'rhs'. That way a lot of diffs will go away. Both names are equally (non-)suggestive.

vasslitvinov replied:
This renaming will make these changes much easier to look at.

e-kayrakli replied:
I agree that `ce` doesn't tell much and I can change it. But `rhs` is confusing to me. Former version of this function was called on a PRIM_MOVE CallExpr, to see if right-hand side needs special treatment. This version takes that primitive itself as an argument, so there is no right-hand side in this context.

mppf replied:
A typical name in the rest of the compiler for a CallExpr would be 'call'.

vasslitvinov commented on compiler/AST/expr.cpp:
```diff
@@ -6039,55 +6052,51 @@ bool CallExpr::codegenPrimMoveRhsIsSpecialPrimop() {
+        ret =  codegenElementPtr(ce->get(1), codegenExprMinusOne(ce->get(2)));
+        retval = true;
       }
 
```
Moving the assignment retval = true into each branch of the conditional above does not make sense to me. Revert this?

e-kayrakli replied:
This seems like a cleanup mistake on my part. These if/else blocks are identical and should probably be removed anyways. Maybe it was initially intended to separate wide/narrow cases, which ended up being the same. I can add a comment to be careful about the distinction.

vasslitvinov commented on compiler/AST/stmt.cpp:
```diff
@@ -928,7 +928,15 @@ CondStmt::codegen() {
 
   if ( outfile ) {
-    info->cStatements.push_back("if (" + codegenValue(condExpr).c + ") ");
+    //here it's very possible that we end up with ( ) around condExpr. It didn't
```
Please add a sentence explaining the benefit of stripping the extra () if they are present.

vasslitvinov commented on test/types/range/elliot/anonymousRangeIter.skipif:
```diff
@@ -10,3 +10,6 @@ COMPOPTS <= --no-inline
 # test greps generated c code, will fail for llvm
 COMPOPTS <= --llvm
+
+# --no-denormalize creates slightly differnet temporary names
```
Are your changes to anonymousRangeIter.expectedGenCode sufficient to handle the slightly differnet temporary names? If they are, skipif should probably remain unchanged.

e-kayrakli replied:
No they are not. I can take a look at changing the test to cover both cases, but it is not obvious to me how I can do that.

vasslitvinov commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,494 @@
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+#define ActualUseDefCastMap std::map<SymExpr*, std::pair<Expr*,Type*> >
```
ActualUseDefCastMap should be a typedef, not #define.

vasslitvinov commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,494 @@
+      }
+
+      std::vector<BaseAST*> asts;
+      collect_asts(fnSym->body, asts);
```
I suggest replacing the above collect_asts with collectSymExprs and moving it inside the if(!cachedGlobalManip) below, to save some compilation time.

vasslitvinov replied:
Ah nevermind I see you need CallExprs in any case.

e-kayrakli commented:
Currently testing new primitive safety check.

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -1139,7 +1140,7 @@ Type* getRefTypesForWideThing(GenRet wide, Type** wideRefTypeOut)
         wideRefType = wide.chplType;
       } else {
-        INT_ASSERT(0);
+        //INT_ASSERT(0);
```
Without this assert, getRefTypesForWideThing returns NULL. But the assert was there to indicate 'wide' was not in fact a wide thing. So it would be better to just not call it at all in that case, and leave the assert in, I think.

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -1435,7 +1436,7 @@ GenRet codegenFieldPtr(
       // lvalue-ptr to one of them.
       INT_ASSERT(isRecord(ct) || isUnion(ct));
-      INT_ASSERT( base.isLVPtr != GEN_VAL );
+      //INT_ASSERT( base.isLVPtr != GEN_VAL );
```
Why comment out this assert? It seems to me codegenFieldPtr can't do it's job if base is a value. What case are you running in to?

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -1477,8 +1478,8 @@ GenRet codegenFieldPtr(
-    if( !fLLVMWideOpt ) INT_ASSERT(base.isLVPtr == GEN_PTR);
-    else INT_ASSERT(base.isLVPtr != GEN_VAL);
+    //if( !fLLVMWideOpt ) INT_ASSERT(base.isLVPtr == GEN_PTR);
+    //else INT_ASSERT(base.isLVPtr != GEN_VAL);
```
I'm starting to wonder if you just commented out all asserts during development? Could these be turned back on?

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -1930,7 +1931,8 @@ GenRet codegenDeref(GenRet r)
   } else {
-    INT_ASSERT(0); // not a reference.
+    //when an svec member value is returned and it is actually an address
+    //INT_ASSERT(0); // not a reference.
```
and this one

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -5858,124 +5885,112 @@ GenRet CallExpr::codegenPrimMove() {
 
-bool CallExpr::codegenPrimMoveRhsIsSpecialPrimop() {
-  CallExpr* rhs    = toCallExpr(get(2));
+static bool codegenIsSpecialPrimitive(BaseAST* target, Expr* e, GenRet& ret) {
```
codegenPrimMoveRhsIsSpecialPrimop had a name that was long enough to describe what the function did, but codegenIsSpecialPrimitive does not. Could you add a comment above this function declaration describing what the function does, what its arguments mean, and in what situations it is called?

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -4103,10 +4105,17 @@ GenRet CallExpr::codegen() {
+        }
       }
 
+      codegenIsSpecialPrimitive(formal, actual, arg);
```
I'd appreciate a comment here that says that codegenIsSpecialPrimitive will replace arg with something else if it takes action.

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -5479,7 +5487,7 @@ GenRet CallExpr::codegenPrimitive() {
       Type* src = get(2)->typeInfo();
 
-      if (dst == src) {
+      if (dst == src && !(is_int_type(dst) || is_uint_type(dst)) ) {
```
This change worries me a little bit. Is the problem actually that Chapel doesn't always have the complete picture in it's type system? If that's the case, should this be checking for extern types? Why should int/uint receive special treatment here?

e-kayrakli replied:
What I needed to have here is the size of C `int` for target architecture. C promotes smaller integers when some arithmetic is run on them. So following two snippets have different behaviors.

  ```
  uint8_t i1 = 255;
  uint8_t i2 = 1;
  uint8_t sum = i1+i2; // promote to int, sum, truncate... sum = 0

  if(sum) {
  }
  else {
    // we'll end up here
  }
  ```

  vs

  ```
  uint8_t i1 = 255;
  uint8_t i2 = 1;

  if(i1+i2) { // promote to int, sum... sum = 256
    // we'll end up here
  }
  else {
  }
  ```

If I could query size of `int` then this if should change to only include smaller-than-int integers. For now, I got stuck with all integers, as I couldn't figure out a simple/elegant way of finding size of `int`.

mppf replied:
Hi - OK, in that case I don't think you're obligated to solve this issue, but at least put a comment explaining the problem.

mppf commented on compiler/AST/expr.cpp:
```diff
@@ -5808,12 +5816,31 @@ GenRet CallExpr::codegenPrimitive() {
+#ifdef HAVE_LLVM
+        codegenStoreLLVM(*specRet, get(1));
+#endif
+      }
```
2 questions:
1) Why do PRIM_GET_REAL and PRIM_GET_IMAG need special treatment here? Why can't that be handled in codegenIsSpecialPrimitive?
2) Even if PRIM_GET_REAL/IMAG need special logic here, couldn't you use codegenAssign either way, and just adjust the arguments to it?

e-kayrakli replied:
I inherited this special treatment from previous helper function directly. I don't know the reason why it was there. Will try to get rid of it.

mppf replied:
If you inherited it from the previous function - don't feel obligated to fix it now - but of course I would be happy if you did fix it. Anyway it would help at least to put a comment indicating that this code doesn't necessarily need to stay here.

mppf commented on compiler/AST/primitive.cpp:
```diff
@@ -157,9 +157,16 @@ returnInfoCast(CallExpr* call) {
+    else if(ct->symbol->hasFlag(FLAG_WIDE_REF)) {
+      return ct->getField(2)->type;
+    }
+  }
```
Good

mppf commented on compiler/AST/primitive.cpp:
```diff
@@ -157,9 +157,16 @@ returnInfoCast(CallExpr* call) {
+    }
+  }
+  INT_FATAL(call, "attempt to get value type of non-reference type");
+  return ct->getField(1)->type; //dummy
```
I'd return NULL instead here for the C type checker.

mppf commented on compiler/AST/stmt.cpp:
```diff
@@ -928,7 +928,16 @@ CondStmt::codegen() {
+    std::string c_condExpr = codegenValue(condExpr).c;
+    if (c_condExpr[0] == '(' && c_condExpr[c_condExpr.size()-1] == ')') {
+      c_condExpr = c_condExpr.substr(1, c_condExpr.length()-2);
+    }
```
I'd make this a while loop.. why stop at removing 1 set of unnecessary parenthesis?

mppf commented:
Does this PR pass full testing + gasnet testing with denormalize turned off? I personally would lean towards making the codegen changes first, in a separate PR, and then adding the denormalize pass in a 2nd PR. I would guess that the codegen changes are more likely to cause problems with various compilers/other configurations.
Alternatively, you could commit the PR as is except for making denormalize off-by-default. Once we deal with any testing problems from the codegen changes, we'd turn denormalize on by default and again deal with any testing problems. I'd just like those to be 2 separate steps.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
```
I don't understand the purpose of globalManipFuncCache. I'd expect you to maintain a map from FnSymbo* -> bool where the bool indicates whether or not a call to that function could modify state (and so should prevent a reordering as is possible with denormalization). I'd like to see that analysis written in a separate C file, even if for now it only works at this point in compilation. I'd personally call it "side effect free" or "pure". I'd expect the code to do this side-effect-free analysis separately from the denormalization, where they seem unnecessarily and confusingly coupled now. (But I could be misunderstanding).

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+  }
+}
+
+void denormalize(FnSymbol *fn) {
```
I'd appreciate a general comment about what denormalization is, what kind of things get denormalized, and why denormalization is useful. You have a lot of that in your PR message but I'd put it in a comment here. It can be very helpful to see a written description of the transformation being performed.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+      //defPar has to be a move without any coercion
+      CallExpr* ce = toCallExpr(defPar);
+      if(ce) {
+        if(ce->isPrimitive(PRIM_MOVE)) {
```
I think this should check for PRIM_MOVE and PRIM_ASSIGN

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+          Type* lhsType = ce->get(1)->typeInfo();
+          Type* rhsType = ce->get(2)->typeInfo();
+          if(lhsType == rhsType) {
+            if(! canPrimMoveCreateCommunication(ce)) {
```
Why does denormalizing care if the def can create communication? Is it just that more work would be required in codegen, or is there something fundamental here?

e-kayrakli replied:
`chpl_gen_comm_get` "returns" the value in a pointer argument. ie

  ```
  int tmp;
  chpl_gen_comm_get(&tmp, ...);
  if(tmp > 0)
    ...
  ```
  Here, `tmp` cannot be denormalized.

mppf replied:
This seems to me to have more to do with pointers and aliasing than communication.

e-kayrakli replied:
Also, call to chpl_gen_comm_get is generated during codegen. ie at denormalize time, it is just a PRIM_MOVE not a function call. That's why I named it this way. Do you suggest I make any changes on this?

mppf replied:
I'd call it something like " primMoveGeneratesCommCall" instead, but the best thing you can do is to explain in a comment above that function what case it is detecting.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+          Expr* defPar = def->parentExpr;
+          if(CallExpr* ceTmp = toCallExpr(defPar)) {
+            if(!isRecord(ceTmp->get(1)->typeInfo())) { //to preserve pass-by-value
+              if(isExprSafeForReorder(def)) {
```
Why does isExprSafeForReorder exist at all? If I had

      var a = someSideEffectingCall();  // e.g. call contains writeln()
      someOtherFunction(a);

then it would be valid to change it to

      someOtherFunction(someSideEffectingCall());

because the C compiler must run someSideEffectingCall() before someOtherFunction().

So, to me, it seems that there is only 1 issue: could the denormalization change the order of execution of the calls?

Consider this example:

      var a = someSideEffectingCall();  // e.g. call contains writeln()
      var b = someSideEffectingCall();  // e.g. call contains writeln()
      someOtherFunction(a,b);

We can change that into

      var a = someSideEffectingCall();  // e.g. call contains writeln()
      someOtherFunction(a, someSideEffectingCall());

without changing the order of the side-effects.

I'll use the terms "reads state" and "modifies state". Generally a function with "side effects" would "modify state". E.g. doing I/O or writing to global or local variables.

So, I'd use these rules for denormalizing:

 * If the definition and the use have any control flow or operations that modify state between them, do not denormalize
 * Otherwise, if the use is the last actual in a function call, it can be denomralized
 * Otherwise it cannot be denormalized

The only way I can think of improving upon these rules is to account for some operations that can always be reordered (and I suspect PRIM_ADDR_OF is that way) but I don't think that improvement is critical right now.

mppf replied:
And I suppose modifications to state in between the def and the use are OK if you can prove that the def and the use couldn't read/write the same state.
Also note that any use of atomic or sync variable should count as "modifies state" to support the memory consistency model.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+              (use, defCastBundle));
+          continue;
+        }
+      }
```
I suspect this "defer actuals" business would not be necessary (and the code would be simpler) if you first decided, for each VarSymbol in a function's body, whether or not to denormalize it, and then, in a separate action, did the denormalizing.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+          if(CallExpr* ceTmp = toCallExpr(defPar)) {
+            if(!isRecord(ceTmp->get(1)->typeInfo())) { //to preserve pass-by-value
+              if(isExprSafeForReorder(def)) {
+                if(!possibleDepInBetween(defPar, usePar)) {
```
This psosibleDepInBetween business is written as a for loop, which means that the analysis part of this algorithm is O(n*2) where n is the number of expressions in a function.

One way to change it to be O(n) is to create a maximum distance between the def and the use (e.g. stop if more than 10 expressions are in-between).

Another way would be to create a count of the number of operations that modify state at each position in the function/block.

E.g. if we have:

    var a = someSideEffectingCall();
    var b = 1;
    var c = 2;
    someSideEffectingCall();

First, record whether or not each statement would inhibit reordering:

    BAD  var a = someSideEffectingCall();
     OK   var b = 1;
     OK   var c = 2;
    BAD  someSideEffectingCall();

Then, assign the value 0 to OK (doesn't inhibit reordering) and 1 to BAD (does inhibit reordering):


     1  var a = someSideEffectingCall();
     0  var b = 1;
     0  var c = 2;
     1  someSideEffectingCall();

Now, compute from these values to a running total:

     1  1 var a = someSideEffectingCall();
     1  0 var b = 1;
     1  0 var c = 2;
     2  1 someSideEffectingCall();

Now, when asking whether or not two statements can be reordered, we can compare this running total to find out if there were any statements in-between that had a side effect. If the running total is equal, there were no such statements. (Of course you have to better define "between" and handle the boundaries correctly).

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,602 @@
+      //
+      //The issue seemed to be yielding string from an iterator
+      if(CallExpr* useParentCe = toCallExpr(usePar)) {
+        if(useParentCe->isPrimitive(PRIM_FTABLE_CALL)) {
```
Is this issue actually from strings (and all records, generally) that have blank argument intent being passed as pointers, but not marked as ref intent at this point?

e-kayrakli commented:
Summary of latest set of commits:

- Expression safety analysis is now in a separate file: `util/exprAnalysis.cpp`
- Temporaries def'd in `PRIM_ASSIGN`s are now denormalized
- Denormalize is now off-by-default: I was thinking about doing it for performance reasons, however I solved the issue. Currently it is off, so we can easily test new codegen changes first before making it on-by-default.
- Minor enhancements
- More comments

Todo:
- Restructure requested by @mppf.
- Tests(new set of commits tested locally with just a few tests)

e-kayrakli commented:
Current version addresses all of the issues to some extent. I will be running different paratest configurations on it today an tomorrow and update the test status. But I think this can go in next week with denormalize off-by-default. Once we are clear of any codegen errors, it can be made on-by-default.

mppf commented on compiler/include/exprAnalysis.h:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Returns true if `e` has no side effects. Checked side effects are:
+ *  - Read/write to a global
+ *  - Is/containsessential primitive
```
I think you're missing a space

mppf commented on compiler/include/exprAnalysis.h:
```diff
@@ -0,0 +1,45 @@
+ * Returns true if `e` has no side effects. Checked side effects are:
+ *  - Read/write to a global
+ *  - Is/containsessential primitive
+ *  - If it's a function call has ref arguments
```
Do you mean "Calls to functions with ref formal arguments" ?
I would think that functions that have 'const ref' formal arguments would be OK.

But, I'd put a comment along the lines of "For now, this is a very conservative analysis. A more precise analysis could distinguish between reads and writes to memory and to take into account alias analysis."

mppf commented on compiler/include/exprAnalysis.h:
```diff
@@ -0,0 +1,45 @@
+
+bool isRegisteredGlobalManip(FnSymbol* fn);
+bool getGlobalManip(FnSymbol* fn);
+void registerGlobalManip(FnSymbol* fn, bool manip);
```
These functions only make sense if the analysis stores a global array. I'd suggest encapsulating the globals in some way - perhaps in a class - so that the analysis could be re-used in different passes.

E.g. 

class SideEffectAnalysis {
  private:
     safeExprCache, globalManipFnCache, etc
  public:
    analyzeFunction(FnSymbol* fn);
    bool exprHasNoSideEffects(Expr* e);
    ...
}

The risk with storing this information in globals is that two passes will use it and mutate the AST and FnSymbols in between. Passes will generally need to be able to control when re-analysis is done.

mppf commented on compiler/util/exprAnalysis.cpp:
```diff
@@ -0,0 +1,241 @@
+ * Returns true if `e` has no side effects. Checked side effects are:
+ *  - Read/write to a global
+ *  - Is/containsessential primitive
+ *  - If it's a function call has ref arguments
```
Update this comment to match the header (which I suggested changes to)

mppf commented on compiler/util/exprAnalysis.cpp:
```diff
@@ -0,0 +1,241 @@
+ */
+bool exprHasNoSideEffects(Expr* e) {
+  if(safeExprCache.count(e) > 0) {
+    return safeExprCache[e];
```
I would think that this safeExprCache is actually only really useful when denormalize is thinking about a particular function (for example). Calls to other functions would be handled by the cache of functions (globalManipFuncCache), right?
That suggests to me an different interface in exprAnalysis.h... e.g. two "analysis" - is a function side effect free? and is an expression side effect free? that each have different state.

mppf commented on compiler/util/exprAnalysis.cpp:
```diff
@@ -0,0 +1,241 @@
+      if(cachedGlobalManip) {
+        if(globalManipFuncCache[fnSym]) {
+          return false;
+        }
```
I'd expect this part to be structured like this:

      if(cachedGlobalManip) {
        bool fnHasSideEffects = globalManipFuncCache[fnSym];
        return !fnHasSideEffects;
      }

Is there some reason you can't write it that way?

mppf commented on compiler/util/exprAnalysis.cpp:
```diff
@@ -0,0 +1,241 @@
+        if(isReferenceType(formal->typeInfo())) {
+          safeExprCache[e] = false;
+          return false;
+        }
```
You'd move this check out of if(cacheGlobalManip) and have it set globalManipFuncCache[fnSym] = true.

mppf commented on compiler/util/exprAnalysis.cpp:
```diff
@@ -0,0 +1,241 @@
+        }
+      }
+      //it shouldn't be touching any globals at this point
+      globalManipFuncCache[fnSym] = false;
```
Generally speaking, I think it'd be nice to put all of this analysis for whether or not a FnSymbol is safe into a separate function. So the above block would become:

      if(cachedGlobalManip) {
        bool fnHasSideEffects = globalManipFuncCache[fnSym];
        return !fnHasSideEffects;
      }

      bool fnHasSideEffects = determineIfFnHasSideEffects(fnSym);
      globalManipFuncCache[fnSym] = fnHasSideEffects;
      return fnHasSideEffects;

mppf commented on compiler/util/exprAnalysis.cpp:
```diff
@@ -0,0 +1,241 @@
+          }
+          e = e->next;
+        }
+        */
```
I think it's reasonable to remove this comment. I don't think you need to worry about the case it's describing with my current understanding of how you're structuring denormalize. However, if you keep it this way, I think it's important for the comments in the exprAnalysis.h to indicate that this analysis does not take into account subexpressions (so PRIM_MOVE, x, call sideEffectingFn() won't even think about what sideEffectingFn does - just what the outermost expr does).

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+ * - Its def is a PRIM_MOVE or PRIM_ASSIGN, with no possible communication
+ * - RHS and LHS are of same type and non-extern
+ *
+ * - Its use a suitable primitive
```
Its use is a

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        registerGlobalManip(fn, true);
+        cachedGlobalManip = true;
```
Isn't this code already in exprAnalysis.cpp ? Did you just forget to remove it from here?

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+    Type* castTo = defCastPair.second;
+
+    // if parent of def is gone, it means it has been denormalized in the
+    // earlier passes
```
If the parent of a def is gone, why do we "defer" it at all? It looks like the loop calling this code only checks the number of "deferred" symbols. So why bother keeping them in a list? Couldn't you just make deferredSyms into a bool?

mppf replied:
I think you explained about deferring earlier ? Or maybe not ? It seems strange/overcomplicated so if it can't be simplified at least put a comment. It's not clear to me why we'd want to stop the analysis only after no defs had a NULL parentExpr.

e-kayrakli replied:
See "On deferring business" comment below.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+      bool isFirstRound = true;
+      do {
+        candidates.clear();
+        deferredSyms.clear();
```
Part of the reason I'm confused about deferredSyms is that you clear it each iteration here.

e-kayrakli replied:
See "On deferring business" comment below.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+}
+
+void findCandidatesInFunc(FnSymbol* fn, Vec<Symbol*> symVec,
+    UseDefCastMap& udcMap) {
```
I'd generally prefer to have different function names for the 2 variations on this function. I think it's easier to find the definition that way.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+      //between use and def
+      if(exprHasNoSideEffects(def)) {
+        if(!unsafeExprInBetween(defPar, usePar)) {
+          std::pair<Expr*, Type*> defCastPair(def, castTo);
```
It would probably be better to use a struct / typedef instead of std::pair here but that is a nit.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+        denormalizeOrDeferCandidates(candidates, deferredSyms);
+
+        isFirstRound = false;
+      } while(deferredSyms.count() > 0);
```
I would have expected this loop to terminate only after denormalization took no action (that is, the number of denormalized symbols was 0).

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+    }
+  } // end loop for symbol
+  if(!cachedGlobalManip) {
+    registerGlobalManip(fn, false);
```
I don't understand this bit... why would a function no longer manipulate globals if you got this far? I'd expect  the exprAnalysis.cpp to independently analyze functions (independent of denormalization) to find if they could be called in a side-effect-free way.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+                  //at this point we now that def is fine
+                  def = ce->get(2);
+
+                  //now check if we need to case it when we move it
```
case it -> cast it

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+                ce->isPrimitive(PRIM_GET_MEMBER_VALUE) ||
+                (ce->isPrimitive(PRIM_MOVE) && 
+                 ce->get(1)->typeInfo() !=
+                 ce->get(2)->typeInfo()))) {
```
I'd write this section like this:

     if (ce->isPrimitive(PRIM_ADDR_OF) ||
         ce->isPrimitive(PRIM_ARRAY_GET) ||
         ... ) {
         // Don't denormalize; type of use unsupported 
      } else {
        // use is OK for denormalization
        use = se;
      }

I generally prefer to use an empty 'if' block with a comment to having complicated logical expressions. But that's just me.

mppf commented on compiler/passes/denormalize.cpp:
```diff
@@ -0,0 +1,457 @@
+      }
+      else {
+        return false;
+      }
```
I think it would be nicer to merge these two else blocks, ie

if (!use || !def) {
  return false;
}

mppf commented:
@e-kayrakli - thanks for updating this PR with my comments. I can see a more understandable version taking shape. Besides the individual comments I made on the code, you need to update your PR message. Now that you've fixed some performance issuess, isn't the compiler faster when running with denormalization on?

mppf commented:
Passed quickstart testing

mppf commented:
Passed quickstart+gasnet testing.

mppf commented:
Passed quickstart+gasnet testing with --denormalize

e-kayrakli commented:
On "deferring" business

NOTE:I used to have deferred _actuals_ in the initial implementations of
the pass, which we don't have anymore. This is a completely different
idea.

Consider following snippet:

  ```
  var t1, t2, t3;
  t1 = f();
  t2 = t1;
  t3 = t2;
  ```

and corresponding pseudo-AST

  ```
  (1 move (2 SymExpr t1), (3 CallExpr f))
  (4 move (5 SymExpr t2), (6 SymExpr t1))
  (7 move (8 SymExpr t3), (9 SymExpr t2))
  ```

In first run we will have a map as follows (use,def expr keypair)

  ```
  use               def
  --------------------------------
  (6 SymExpr t1)    (3 CallExpr f)
  (9 SymExpr t2)    (6 SymExpr t1)
  ```

Than in `denormalizeOrDeferCandidates`, assume we denormalize (t1,f)
pair first, which will replace node 6 with 3 and remove 1 altogether. At
that point when we try to denormalize the second pair(t2,t1) in our
list, node id 6 will still be in the map but it's parent pointer will be
`NULL`, since it's removed by `6->replace(3)` operation in the previous
iteration.

[I might be leaking compiler memory in the logic I just described]

At that point, we still know that t2 is a good candidate for removal,
but its def has changed, and we need to reanalyze for safety. So we add
t2 to deferredSyms for future iterations. We can get rid of
deferredSyms, but that would mean reiterating all symbols in the
function. And I am _almost_ certain that we can never generate new
candidates once we start denormalizing. So, generating candidates once,
and iterating over and over until we don't defer anything(thus
`do{}while(deferredSyms.count()>0)`) seemed reasonable to me.

Is there anything that I am overlooking? Is there a clearer way of
implementing this? Or; should I explain this in a comment? Maybe my
little comment "if parent of def is gone, it means it has been
denormalized in the earlier passes" is more misleading than helpful

benharsh commented:
@e-kayrakli : I'm seeing a compile-time assertion error in our compiler for the following primers on master (only with the --denormalize flag thrown):
- arrays.chpl
- associative.chpl

The error looks like this for arrays.chpl:
```$CHPL_HOME/modules/internal/DefaultRectangular.chpl:1571: internal error: assertion error [expr.cpp:5400]```

This is on OSx with clang. Can you repro?

e-kayrakli commented:
@benharsh Yes, I can on my linux with gnu. I gave it a quick look and I think that part where the assertion is triggered is relatively new. It's weird that Michael didn't hit that error as I think he was running tests with this PR merged to master. So, maybe I am wrong. But in general I was never seeing PRIM_CHPL_COMM* primitives at that stage of compilation. Which may have changed recently(Especially I don't remember strided primitives at all)

It looks like I have to add protection for those primitives in the pass, shouldn't be too hard. Did you happen to see any other issue with --denormalize flag?

e-kayrakli commented:
I have sent a fix to @benharsh .
